custom/icds_reports/ucr/expressions.py

```diff
@@ -215,15 +215,64 @@ def _get_case_form_ids(self, case_id, context):
         context.set_cache_value(cache_key, xform_ids)
         return xform_ids
 
-    def _get_form_json(self, form, context):
-        cache_key = (XFORM_CACHE_KEY_PREFIX, form.get_id)
-        if context.get_cache_value(cache_key) is not None:
-            return context.get_cache_value(cache_key)
+    @staticmethod
+    def _get_form_json_list(case_id, xforms, context, domain):
+        domain_filtered_forms = [f for f in xforms if f.domain == domain]
+        return [FormsInDateExpressionSpec._get_form_json(f, context) for f in domain_filtered_forms]
+
+    @staticmethod
+    def _get_form_json(form, context):
+        cached_form = FormsInDateExpressionSpec._get_cached_form_json(form, context)
+        if cached_form is not None:
+            return cached_form
 
         form_json = form.to_json()
-        context.set_cache_value(cache_key, form_json)
+        FormsInDateExpressionSpec._set_cached_form_json(form, form_json, context)
         return form_json
 
+    @staticmethod
+    def _bulk_get_form_json_from_es(forms):
+        form_ids = [form.form_id for form in forms]
+        es_forms = FormsInDateExpressionSpec._bulk_get_forms_from_elasticsearch(form_ids, source=True)
+        for f in es_forms:
+            # for parity with what comes out of Riak
+            del f['timeEnd']
+        return {
+            f['_id']: f for f in es_forms
+        }
+
+    @staticmethod
+    def _get_cached_form_json(form, context):
+        return context.get_cache_value(FormsInDateExpressionSpec._get_form_json_cache_key(form))
+
+    @staticmethod
+    def _set_cached_form_json(form, form_json, context):
+        context.set_cache_value(FormsInDateExpressionSpec._get_form_json_cache_key(form), form_json)
+
+    @staticmethod
+    def _get_form_json_cache_key(form):
+        return (XFORM_CACHE_KEY_PREFIX, form.form_id)
+
+    @staticmethod
+    def _bulk_get_forms_from_elasticsearch(form_ids, source):
+        forms = mget_query('forms', form_ids, source)
```

Comment:
Hey @czue Looks like this expression `icds_get_case_forms_in_date` is used only in one ICDS UCR with `count: True`. Was that not the case when you wrote this? I am thinking to make a PR to just get the counts and meta fields explicitly and remove full form loading. 

Also was the performance gain over shard DBs or over the blobdb that's needed for loading the full form?

Reply:
Hard to remember offhand, but I'm pretty sure it was more broadly used before that. Should be trivial to checkout the last commit of this PR and grep again I would think.

Performance was relative to blobdb.

@emord might remember why we were doing this in memory versus pushing the query into ES on the first go around. I think it had something to do with caching and re-use.

Reply:
But sorry, just to be more clear:

> I am thinking to make a PR to just get the counts and meta fields explicitly and remove full form loading.

I don't see any reason why this wouldn't be a good idea at this point, but @emord might.

Reply:
These are used much less than they were when initially, however as pointed out in the scale meeting these are still used for more than just counts https://github.com/dimagi/commcare-hq/blob/dcaf9aebd6f22d09b002e53f8bc5ea75e77d0f54/custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json#L1830-L1857

Cory is right that we originally didn't do rely on ES counts so that the these cached queries could be reused.

(FWIW I think your idea to create a new expression to pull these from a UCR data source is pretty good. It is not an ideal solution, but it sounds like it hits a good balance of LOE vs better place than we are now)